include/pybind11/detail/internals.h

```diff
@@ -280,21 +281,112 @@ inline internals **&get_internals_pp() {
     return internals_pp;
 }
 
+#if PY_VERSION_HEX >= 0x03030000
+inline bool raise_err(PyObject *exc_type, const char *msg) {
+    bool err_occurred = PyErr_Occurred() != nullptr;
+    if (err_occurred) {
+        raise_from(exc_type, msg);
+    } else {
+        PyErr_SetString(exc_type, msg);
+    }
+    return err_occurred;
+};
+
+// forward decl
+inline void translate_exception(std::exception_ptr);
```

Comment:
Forward declarations should never have `inline`:

https://stackoverflow.com/questions/9317473/forward-declaration-of-inline-functions

(I find this highly counter-intuitive myself.)

Reply:
C99 requires the inline on the definition; are you sure C++ cares? Looking at the spec, it seems clearly allowed unless it's nested inside another function https://en.cppreference.com/w/cpp/language/inline & https://en.cppreference.com/w/cpp/language/declarations#Specifiers clearly seem to indicate it's allowed, and I didn't have any problems on any complier I quickly checked on godbolt.

Reply:
> C99 requires the inline on the definition;

Ouch.

My previous comment was based on what I remembered from working on
https://github.com/pybind/pybind11/pull/3179.

From memory, in a Google-internal chat I was advised to not use `inline` in forward declarations.

Am I sure? No. I was asking for advice and then going with that. The work on PR #3179 left me believing that even compilers are unusually confused about what is correct.

To be pragmatic, we can reduce it to: do we want to be consistent within pybind11? PR #3179 removed some `inline` based on the advice I got, and to be consistent.

Reply:
I intentionally kept #3179 limited to functions involving `PYBIND11_NOINLINE`. I didn't look around too much, as I very often do, to not let a good project die the death of a thousand cuts.

For this PR, idk. If we're inconsistent already and the advice I got in Aug last year isn't universally accepted (as I was thinking until now) ... shrug :-)